src/paths/bellman_ford.c

```diff
@@ -203,3 +210,285 @@ int igraph_shortest_paths_bellman_ford(const igraph_t *graph,
+ *          \cli IGRAPH_EINVVID
+ *           \p from is invalid vertex id, or the length of \p to is
+ *           not the same as the length of \p vertices or \p edges.
+ *          \cli IGRAPH_ENEGLOOP
+ *          Bellman-ford algorithm encounted a negative loop.
+ *
+ * Time complexity: O(|E|*|V|), where |V| is the number of
+ * vertices, |E| the number of edges.
+ *
+ * \sa \ref igraph_shortest_paths() for a faster unweighted version
+ * or \ref igraph_shortest_paths_dijkstra() if you do not have negative
+ * edge weights.
+ *
+ * \example examples/simple/igraph_get_shortest_paths_bellman_ford.c
+ */
+
+int igraph_get_shortest_paths_bellman_ford(const igraph_t *graph,
+                                        igraph_vector_ptr_t *vertices,
+                                        igraph_vector_ptr_t *edges,
+                                        const igraph_integer_t from,
+                                        igraph_vs_t to,
+                                        const igraph_vector_t *weights,
+                                        igraph_neimode_t mode,
+                                        igraph_vector_long_t *predecessors,
+                                        igraph_vector_long_t *inbound_edges) {
+    long int no_of_nodes = igraph_vcount(graph);
+    long int no_of_edges = igraph_ecount(graph);
+    long int *parents;
+    igraph_lazy_inclist_t inclist;
+    long int i, j, k;
+    long int no_of_to;
+    igraph_dqueue_t Q;
+    igraph_vector_t clean_vertices;
+    igraph_vector_t num_queued;
+    igraph_vit_t tovit;
+    igraph_real_t my_infinity = IGRAPH_INFINITY;
+    igraph_bool_t all_to;
+    igraph_vector_t dist;
+
+    if (!weights) {
+        return  igraph_get_shortest_paths(graph, vertices, edges, from, to, mode,
+                                         predecessors, inbound_edges);
+    }
+
+    if (igraph_vector_size(weights) != no_of_edges) {
+        IGRAPH_ERROR("Weight vector length does not match", IGRAPH_EINVAL);
+    }
+
+    IGRAPH_DQUEUE_INIT_FINALLY(&Q, no_of_nodes);
+    IGRAPH_VECTOR_INIT_FINALLY(&clean_vertices, no_of_nodes);
+    IGRAPH_VECTOR_INIT_FINALLY(&num_queued, no_of_nodes);
+    IGRAPH_CHECK(igraph_lazy_inclist_init(graph, &inclist, mode));
+    IGRAPH_FINALLY(igraph_lazy_inclist_destroy, &inclist);
+
+    if ( (all_to = igraph_vs_is_all(&to)) ) {
+        no_of_to = no_of_nodes;
+    } else {
+        IGRAPH_CHECK(igraph_vit_create(graph, to, &tovit));
+        IGRAPH_FINALLY(igraph_vit_destroy, &tovit);
+        no_of_to = IGRAPH_VIT_SIZE(tovit);
+    }
+
+    if (vertices && IGRAPH_VIT_SIZE(tovit) != igraph_vector_ptr_size(vertices)) {
+        IGRAPH_ERROR("Size of `vertices' and `to' should match", IGRAPH_EINVAL);
+    }
+    if (edges && IGRAPH_VIT_SIZE(tovit) != igraph_vector_ptr_size(edges)) {
+        IGRAPH_ERROR("Size of `edges' and `to' should match", IGRAPH_EINVAL);
+    }
+
+    parents = igraph_Calloc(no_of_nodes, long int);
```

Comment:
It looks like this is never freed, that's another leak.

Reply:
changed - added free at the end of the program.